File: src/utils/GeminiProvider.js
/**
 * GeminiProvider - Google Gemini API provider
 * 
 * Features:
 * - Fast embeddings (text-embedding-004)
 * - Strong multilingual support
 * - Generous free tier
 * 
 * Environment:
 *   GEMINI_API_KEY=your_api_key
 *   GEMINI_EMBEDDING_MODEL=text-embedding-004
 *   GEMINI_CHAT_MODEL=gemini-1.5-flash
 * 
 * Get API key: https://aistudio.google.com/app/apikey
 */

import { IntelligentProvider } from '../interfaces/EmbeddingProvider.js';

export class GeminiProvider extends IntelligentProvider {
  constructor(options = {}) {
    super({
      name: 'Gemini',
      dimensions: options.dimensions || 768,
      ...options
    });
    
    this.apiKey = options.apiKey || process.env.GEMINI_API_KEY;
    this.embeddingModel = options.embeddingModel || process.env.GEMINI_EMBEDDING_MODEL || 'text-embedding-004';
    this.chatModel = options.chatModel || process.env.GEMINI_CHAT_MODEL || 'gemini-1.5-flash';
    this.baseUrl = 'https://generativelanguage.googleapis.com/v1beta';
  }

  isAvailable() {
    return !!this.apiKey;
  }

  async embed(text) {
    if (!this.apiKey) return null;

    try {
      const response = await fetch(
        `${this.baseUrl}/models/${this.embeddingModel}:embedContent?key=${this.apiKey}`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            model: `models/${this.embeddingModel}`,
            content: {
              parts: [{ text: text.substring(0, 10000) }]
            }
          })
        }
      );

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Gemini embedding error: ${error}`);
      }

      const data = await response.json();
      return data.embedding?.values || null;
    } catch (err) {
      console.error(`[GeminiProvider] Embedding failed: ${err.message}`);
      return null;
    }
  }

  async embedBatch(texts) {
    if (!this.apiKey || texts.length === 0) {
      return texts.map(() => null);
    }

    try {
      const response = await fetch(
        `${this.baseUrl}/models/${this.embeddingModel}:batchEmbedContents?key=${this.apiKey}`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            requests: texts.map(text => ({
              model: `models/${this.embeddingModel}`,
              content: {
                parts: [{ text: text.substring(0, 10000) }]
              }
            }))
          })
        }
      );

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Gemini batch embedding error: ${error}`);
      }

      const data = await response.json();
      return data.embeddings?.map(e => e.values) || texts.map(() => null);
    } catch (err) {
      console.error(`[GeminiProvider] Batch embedding failed: ${err.message}`);
      // Fallback to individual calls
      return Promise.all(texts.map(t => this.embed(t)));
    }
  }

  async autoTag(key, content) {
    if (!this.apiKey) return [];

    const prompt = `Generate 5-10 relevant tags for this content.
Return ONLY a JSON array of lowercase, hyphenated tags.

Title: ${key}
Content: ${content.substring(0, 2000)}`;

    try {
      const result = await this.generateContent(prompt);
      const cleaned = result
        .replace(/```json/g, '')
        .replace(/```/g, '')
        .trim();
      return JSON.parse(cleaned);
    } catch (err) {
      console.error(`[GeminiProvider] Auto-tag failed: ${err.message}`);
      return [];
    }
  }

  async synthesize(content, task, style = '') {
    if (!this.apiKey) {
      return content.substring(0, 1500);
    }

    const styleGuides = {
      beginner: 'Explain simply for beginners.',
      detailed: 'Be comprehensive with examples.',
      concise: 'Be extremely brief.'
    };

    const prompt = `${styleGuides[style] || ''}

Context:
${content.substring(0, 7000)}

Task: ${task}`;

    try {
      return await this.generateContent(prompt);
    } catch (err) {
      console.error(`[GeminiProvider] Synthesis failed: ${err.message}`);
      return content.substring(0, 1500);
    }
  }

  async extractGraphEntities(content) {
    if (!this.apiKey) return { concepts: [], relationships: [] };

    const prompt = `Extract key concepts and relationships from this text.
Return ONLY JSON in this exact format:
{
  "concepts": ["concept1", "concept2"],
  "relationships": [{"from": "c1", "to": "c2", "type": "related_to", "strength": 0.8}]
}

Text: ${content.substring(0, 3000)}`;

    try {
      const result = await this.generateContent(prompt);
      const cleaned = result
        .replace(/```json/g, '')
        .replace(/```/g, '')
        .trim();
      const parsed = JSON.parse(cleaned);
      return {
        concepts: Array.isArray(parsed.concepts) ? parsed.concepts : [],
        relationships: Array.isArray(parsed.relationships) ? parsed.relationships : []
      };
    } catch (err) {
      console.error(`[GeminiProvider] Graph extraction failed: ${err.message}`);
      return { concepts: [], relationships: [] };
    }
  }

  async generateContent(prompt) {
    const response = await fetch(
      `${this.baseUrl}/models/${this.chatModel}:generateContent?key=${this.apiKey}`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          contents: [{
            parts: [{ text: prompt }]
          }],
          generationConfig: {
            temperature: 0.2,
            maxOutputTokens: 2048
          }
        })
      }
    );

    if (!response.ok) {
      throw new Error(await response.text());
    }

    const data = await response.json();
    return data.candidates?.[0]?.content?.parts?.[0]?.text || '';
  }

  getConfig() {
    return {
      type: 'gemini',
      embeddingModel: this.embeddingModel,
      chatModel: this.chatModel,
      dimensions: this.dimensions,
      hasApiKey: !!this.apiKey
    };
  }
}

File: src/adaptive_memory_engine/providers/gemini_provider.py
"""Google Gemini provider."""

from __future__ import annotations

import json
import logging
import re
from collections.abc import Iterable

import httpx

from .base import IntelligentProvider
from .http import request_with_retry

log = logging.getLogger(__name__)


class GeminiProvider(IntelligentProvider):
    name = "gemini"
    dimensions = 768

    def __init__(
        self,
        api_key: str | None,
        embedding_model: str = "gemini-embedding-001",
        chat_model: str = "gemini-2.5-flash",
        base_url: str = "https://generativelanguage.googleapis.com/v1beta",
        timeout: float = 60.0,
    ) -> None:
        self.api_key = api_key
        self.embedding_model = embedding_model
        self.chat_model = chat_model
        self.base_url = base_url.rstrip("/")
        self._client = httpx.Client(timeout=timeout)

    def is_available(self) -> bool:
        return bool(self.api_key)

    def embed(self, text: str) -> list[float]:
        url = f"{self.base_url}/models/{self.embedding_model}:embedContent"
        r = request_with_retry(
            self._client,
            "POST",
            url,
            headers={"Content-Type": "application/json", "x-goog-api-key": self.api_key or ""},
            json={
                "content": {"parts": [{"text": text[:8000]}]},
                "outputDimensionality": self.dimensions,
            },
        )
        r.raise_for_status()
        return r.json()["embedding"]["values"]

    def embed_batch(self, texts: Iterable[str]) -> list[list[float]]:
        url = f"{self.base_url}/models/{self.embedding_model}:batchEmbedContents"
        reqs = [
            {
                "model": f"models/{self.embedding_model}",
                "content": {"parts": [{"text": t[:8000]}]},
                "outputDimensionality": self.dimensions,
            }
            for t in texts
        ]
        r = request_with_retry(
            self._client,
            "POST",
            url,
            headers={"Content-Type": "application/json", "x-goog-api-key": self.api_key or ""},
            json={"requests": reqs},
        )
        r.raise_for_status()
        return [e["values"] for e in r.json()["embeddings"]]

    def auto_tag(self, key: str, content: str) -> list[str]:
        prompt = (
            "Generate 3-5 concise lowercase tags as JSON array of strings.\n\n"
            f"Key: {key}\nContent: {content[:1500]}\nTags:"
        )
        text = self._generate(prompt)
        return self._parse_tag_list(text)

    def synthesize(self, content: str, task: str, style: str = "concise") -> str:
        prompt = f"Task: {task}\nStyle: {style}\n\nContent:\n{content[:6000]}\n\nResponse:"
        return self._generate(prompt)

    def expand_query(self, query: str) -> list[str]:
        prompt = f"Generate 3 paraphrases as JSON array of strings.\n\nQuery: {query}\nParaphrases:"
        text = self._generate(prompt)
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("queries") or arr.get("paraphrases") or []
            if isinstance(arr, list):
                return [str(x) for x in arr]
        except Exception:  # noqa: BLE001
            pass
        return [query]

    def get_config(self) -> dict:
        return {
            "type": "gemini",
            "embeddingModel": self.embedding_model,
            "chatModel": self.chat_model,
        }

    def _generate(self, prompt: str) -> str:
        url = f"{self.base_url}/models/{self.chat_model}:generateContent"
        r = request_with_retry(
            self._client,
            "POST",
            url,
            headers={"Content-Type": "application/json", "x-goog-api-key": self.api_key or ""},
            json={
                "contents": [{"parts": [{"text": prompt}]}],
                "generationConfig": {"temperature": 0.2},
            },
        )
        r.raise_for_status()
        data = r.json()
        return data["candidates"][0]["content"]["parts"][0]["text"]

    def close(self) -> None:
        self._client.close()

    def _parse_tag_list(self, text: str) -> list[str]:
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("tags", [])
            if isinstance(arr, list):
                return [str(t).strip().lower() for t in arr if str(t).strip()][:10]
        except Exception:  # noqa: BLE001
            pass
        return [m.group(1).lower() for m in re.finditer(r'"([^"]+)"', text)][:10]
